# HG changeset patch
# User Rik <rik@octave.org>
# Date 1546541411 28800
#      Thu Jan 03 10:50:11 2019 -0800
# Branch stable
# Node ID 31b0761cd641ff6038818c0ab45000d3ba3d2149
# Parent  b352120504bc165d41cb12ab7e0ad2d689a1c287
gl-render.cc: Fix static analyzer detected issues (bug #55347).

* gl-render.cc (opengl_texture& operator =, vertex_data& operator =):
Check object to be assigned against this pointer before doing assignment.

diff --git a/libinterp/corefcn/gl-render.cc b/libinterp/corefcn/gl-render.cc
--- a/libinterp/corefcn/gl-render.cc
+++ b/libinterp/corefcn/gl-render.cc
@@ -147,21 +147,24 @@ namespace octave
     ~opengl_texture (void)
     {
       if (--rep->count == 0)
         delete rep;
     }
 
     opengl_texture& operator = (const opengl_texture& tx)
     {
-      if (--rep->count == 0)
-        delete rep;
-
-      rep = tx.rep;
-      rep->count++;
+      if (&tx != this)
+        {
+          if (--rep->count == 0)
+            delete rep;
+
+          rep = tx.rep;
+          rep->count++;
+        }
 
       return *this;
     }
 
     static opengl_texture create (opengl_functions& glfcns,
                                   const octave_value& data);
 
     void bind (int mode = GL_TEXTURE_2D) const
@@ -424,21 +427,24 @@ namespace octave
     ~vertex_data (void)
     {
       if (--rep->count == 0)
         delete rep;
     }
 
     vertex_data& operator = (const vertex_data& v)
     {
-      if (--rep->count == 0)
-        delete rep;
-
-      rep = v.rep;
-      rep->count++;
+      if (&v != this)
+        {
+          if (--rep->count == 0)
+            delete rep;
+
+          rep = v.rep;
+          rep->count++;
+        }
 
       return *this;
     }
 
     vertex_data_rep * get_rep (void) const { return rep; }
   };
 
   class
